README.md
# Fastify MS Graph Change Notification Webhook Plugin

Fastify plugin implementing [MS Graph Change Notifications webhook](https://learn.microsoft.com/it-it/graph/change-notifications-delivery-webhooks?tabs=http).

## Install

```
npm i --save fastify-msgraph-change-notification-webhook
```

## Usage

The plugin requires at least `fastify:4.x`.

```js
const fastifyChangeNotificationWebhook = require('fastify-msgraph-change-notification-webhook')
const fastify = require('fastify')()

/** ... setup fastify instance ... */

fastify.register(fastifyChangeNotificationWebhook, {
    path: '/your-webhook-path'
    subscriptions: {
        'subscription-1': {
            clientState: 'subscription-1-secret',
            callback: (notification) => console.log('subscription-1 sent notification', notification)
        },
        'subscription-2': {
            clientState: 'subscription-2-secret',
            callback: (notification) => console.log('subscription-2 sent notification', notification)
        }
    },
})
```

## Reference

This plugin exposes a `POST` route under the path provided by the plugin which:

* implements [validation url procedure](https://learn.microsoft.com/it-it/graph/change-notifications-delivery-webhooks?tabs=http#notificationurl-validation);
* for each [change notification](https://learn.microsoft.com/it-it/graph/change-notifications-delivery-webhooks?tabs=http#receive-notifications) received: 
  * verifies that the change notification secret is valid;
  * executes a callback.

  Secrets and callbacks are provided to the plugin by the `subscriptions` object. It's a key-value pair JSON where each entry has its own MS Graph `subscriptionId` and contains the following information:
    * `clientState`: the value provided during subscription to validate change notifications;
    * `callback`: a function that will process the notification

The route responds with status code 200 if all callbacks during a request are executed correctly, otherwise a status code 500 will be returned: this means that MS Graph will try to re-process the notification later.

Also, the plugin exposes a copy of the `subscriptions` object through your `fastify` instance.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2023-12-04

## Added

* [`/POST`] webhook
* fastify-plugin

tests/index.test.js
const tap = require('tap')
const sinon = require('sinon')
const Fastify = require('fastify')
const fastifyChangeNotificationWebhook = require('..')


tap.test('change notification plugin', async t => {
    
	const CLIENT_STATE = 'my-state'
	const SUBSCRIPTION_ID = 'my-id'  

	const OK_NOTIFICATION = {
		'id': 'lsgTZMr9KwAAA',
		'subscriptionId': SUBSCRIPTION_ID,
		'subscriptionExpirationDateTime':'2016-03-19T22:11:09.952Z',
		'clientState': CLIENT_STATE,
		'changeType':'created',
		'resource':'users/{user_guid}@{tenant_guid}/messages/{long_id_string}',
		'tenantId': '84bd8158-6d4d-4958-8b9f-9d6445542f95',
		'resourceData':
        {
        	'@odata.type':'#Microsoft.Graph.Message',
        	'@odata.id':'Users/{user_guid}@{tenant_guid}/Messages/{long_id_string}',
        	'@odata.etag':'W/"CQAAABYAAADkrWGo7bouTKlsgTZMr9KwAAAUWRHf"',
        	'id':'{long_id_string}'
        }
	}

	const TEST_NOTIFICATION = {
		'value': [
			OK_NOTIFICATION,
			{
				'id': 'lsgTZMr9KwAAA',
				'subscriptionId': 'wrong-subscrption-id',
				'subscriptionExpirationDateTime':'2016-03-19T22:11:09.952Z',
				'clientState': CLIENT_STATE,
				'changeType':'created',
				'resource':'users/{user_guid}@{tenant_guid}/messages/{long_id_string}',
				'tenantId': '84bd8158-6d4d-4958-8b9f-9d6445542f95',
				'resourceData':
            {
            	'@odata.type':'#Microsoft.Graph.Message',
            	'@odata.id':'Users/{user_guid}@{tenant_guid}/Messages/{long_id_string}',
            	'@odata.etag':'W/"CQAAABYAAADkrWGo7bouTKlsgTZMr9KwAAAUWRHf"',
            	'id':'{long_id_string}'
            }
			},
			{
				'id': 'lsgTZMr9KwAAA',
				'subscriptionId': SUBSCRIPTION_ID,
				'subscriptionExpirationDateTime':'2016-03-19T22:11:09.952Z',
				'clientState': 'wrong-client-state',
				'changeType':'created',
				'resource':'users/{user_guid}@{tenant_guid}/messages/{long_id_string}',
				'tenantId': '84bd8158-6d4d-4958-8b9f-9d6445542f95',
				'resourceData':
            {
            	'@odata.type':'#Microsoft.Graph.Message',
            	'@odata.id':'Users/{user_guid}@{tenant_guid}/Messages/{long_id_string}',
            	'@odata.etag':'W/"CQAAABYAAADkrWGo7bouTKlsgTZMr9KwAAAUWRHf"',
            	'id':'{long_id_string}'
            }
			}
		]
	}

	t.test('plugin ok', async t => {
		const fastify = Fastify()
		const BASE_CLIENT_SUBSCRIPTIONS = {}
		const mockCallback = sinon.fake.resolves()
		BASE_CLIENT_SUBSCRIPTIONS[SUBSCRIPTION_ID] = {
			clientState: CLIENT_STATE,
			callback: mockCallback
		} 
        
		await fastify.register(fastifyChangeNotificationWebhook, {
			path: '/',
			subscriptions: BASE_CLIENT_SUBSCRIPTIONS,
		})
		await fastify.ready()
		t.strictSame(fastify.subscriptions, BASE_CLIENT_SUBSCRIPTIONS)

		t.test('webhook returns query validation token', async t => {    
			const validationToken = 'testValidationToken'
			const { statusCode, payload } = await fastify.inject({
				path: '/',
				method: 'POST',
				query: {
					validationToken,
				}
			})
			t.strictSame(statusCode, 200)
			t.strictSame(payload, validationToken)
		})

		t.test('webhook executes only callback from allowed notification', async t => {    
			const { statusCode } = await fastify.inject({
				path: '/',
				method: 'POST',
				body: TEST_NOTIFICATION,
			})
			t.strictSame(statusCode, 200)
			t.ok(mockCallback.calledOnceWith(fastify.log, OK_NOTIFICATION))
		})

		t.teardown(async() => fastify.close())
	})
    

	t.test('plugin ko', async t => {
		const BASE_CLIENT_SUBSCRIPTIONS = {}
		const mockFailCallback = sinon.fake.rejects()
		BASE_CLIENT_SUBSCRIPTIONS[SUBSCRIPTION_ID] = {
			clientState: CLIENT_STATE,
			callback: mockFailCallback
		} 
		t.test('webhook returns 500 if callback fails', async t => {   
			const fastify = Fastify()
            
			await fastify.register(fastifyChangeNotificationWebhook, {
				path: '/',
				subscriptions: BASE_CLIENT_SUBSCRIPTIONS,
			})
			await fastify.ready() 
			const { statusCode } = await fastify.inject({
				path: '/',
				method: 'POST',
				body: TEST_NOTIFICATION,
			})
			await fastify.close()
			t.strictSame(statusCode, 500)
			t.ok(mockFailCallback.calledOnceWith(fastify.log, OK_NOTIFICATION))
		})

		t.test('no path has been provided', async t => {
			const fastify = Fastify()
            
			try {
				await fastify.register(fastifyChangeNotificationWebhook, {
					subscriptions: BASE_CLIENT_SUBSCRIPTIONS,
				})
				t.fail('test should fail')
			}catch(error) {
				t.ok(error)
			} finally {
				await fastify.close()
			}        
		})

		t.test('invalid subscriptions structure', async t => {
			const fastify = Fastify()
            
			try {
				await fastify.register(fastifyChangeNotificationWebhook, {
					path: '/',
					subscriptions: { 
						wrongSubscription: {
						}
					},
				})
				t.fail('test should fail')
			}catch(error) {
				t.ok(error)
			} finally {
				await fastify.close()
			}      
		})
        
	})
})

types/index.d.ts
import type { FastifyPluginAsync } from 'fastify';
import { AsyncFunction } from 'fastify/types/instance';

declare module 'fastify' {
  interface FastifyInstance {
    /**
     * MS Graph subscriptions
     */
    subscriptions: Record<string, changeNotificationWebhook.Subscription>;
  }
}

type ChangeNotificationWebhook = FastifyPluginAsync<changeNotificationWebhook.ChangeNotificationWebhookOptions>;

declare namespace changeNotificationWebhook {

  export interface Subscription {
    /**
     * secret that will be used to validate notifications
     */
    clientState: string,
    /**
     * callback that will be executed for each notification received
     */
    callback: AsyncFunction
  }

  export interface ChangeNotificationWebhookOptions {
    /**
     * Path where the webhook will be exposed
     */
    path: string,

    /**
     * MS Graph Subscriptions 
     * @default {{}}
     */
    subscriptions: Record<string, Subscription>;
  }

  export const changeNotificationWebhook: ChangeNotificationWebhook
  export { changeNotificationWebhook as default }
}

declare function changeNotificationWebhook(...params: Parameters<ChangeNotificationWebhook>): ReturnType<ChangeNotificationWebhook>
export = changeNotificationWebhook

index.js
const fastifyPlugin = require('fastify-plugin')

function getAllowedNotificationsFromSubscriptions(subscriptions, notifications) {
	return notifications.filter((notification) => {
		const { subscriptionId, clientState } = notification
		const clientSecret = subscriptions[subscriptionId]?.clientState
		return clientSecret && clientSecret === clientState
	})
}

function checkOptions(options) {
	const { path, subscriptions } = options
	
	if(!path && typeof path !== 'string') {
		return new Error('no path has been provided')
	}

	let error = null
	for(const subscriptionId of Object.keys(subscriptions)) {
		const { clientState, callback } = subscriptions[subscriptionId]
		if(!clientState || !callback) {
			error = new Error('subscription '+subscriptionId+' has no valid structure.')
		}
	}
	return error
}
/**
 * 
 * @param {import('fastify').FastifyInstance} fastify 
 * @param {import('./types').changeNotificationWebhookOptions} options 
 */
async function changeNotificationWebhook(fastify, options, next) {
	const error = checkOptions(options)
    
	if(error) {
		next(error)
	}
    
	const {
		subscriptions,
		path,
	} = options
    
	fastify.decorate('subscriptions', { ...subscriptions })
	fastify.post(path, async function(request, reply) {
		const { query, body, log } = request
		const response = {
			status: 200
		}
		if(query.validationToken) {
			return reply.type('text/plain').send(query.validationToken)
		}
		const { value: notifications } = body
		log.debug({ body }, 'received notification')
		const allowedNotifications = getAllowedNotificationsFromSubscriptions(subscriptions, notifications)
		try {
			for await( const notification of allowedNotifications) {
				const { callback } = subscriptions[notification.subscriptionId]
				await callback(log.child({ subscriptionId: notification.subscriptionId }), notification)
			}
		} catch(error) {
			log.error(error, 'error occured during change notification')
			response.status = 500
		} finally {
			reply.status(response.status).send()
		} 
	})

	next()
}

module.exports = fastifyPlugin(changeNotificationWebhook, {
	fastify: '4.x',
	name: 'fastify-msgraph-change-notification-webhook'
})